### PR TITLE
Extend the UserProfileCard and GroupProfileCard to allow for adding links 

### PR DESCRIPTION
**Pull request by jrwpatterson**
## Hey, I just made a Pull Request!

The group and user profile cards don't display much information about the relative users, in our system we would like to add more profile information, this update changes the type definition for `GroupEntity` and `UserEntity` to add the ability for any field to be added to profile in as a `Record<string, string>`.

Then in both the `GroupProfileCard` and the `UserProfileCard` it adds functionality so that if either the entities Links or extra profile fields are added to then the fields will get added to the card in a graceful way. 

That way if we want to add extra data to the catalog user, eg slack link, job title, contact information, timezone (we are international so good to see what time zone the user is) to the user we can in an easy way. Same goes for group, if we were to add for example group slack, or extra data like where the group sits in the office, or links of the work the group is involved in we can. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Screen Shot 2022-10-14 at 2 44 58 pm

Screen Shot 2022-10-14 at 2 44 53 pm

### CODE REVIEW

**Review by samjmarshall — APPROVED** [review]

**Comment by djamaile** [comment]
Hi! you forgot to build the api-reports. From the root you can run `yarn build:api-reports plugins/<changed-plugin>`.

File: ADOPTERS.md
Comment (awanlin) [inline ADOPTERS.md:218]: Maybe this could go into it's own PR as it's not directly related?

File: storybook/.storybook/preview-head.html
Comment (awanlin) [inline storybook/.storybook/preview-head.html:3]: There are MUI icons being used elsewhere in the Storybook why does this need to be added?

File: packages/core-components/src/components/Table/Table.tsx
Comment (awanlin) [inline packages/core-components/src/components/Table/Table.tsx:376]: Just curious but why is this change needed? I might not seeing that some of the other changes might need it

File: .changeset/tender-jeans-clean.md
Comment (awanlin) [inline .changeset/tender-jeans-clean.md:1]: Can you please include an example of how this will be used? So examples of how to use the card and then how to add the extra properties as well. This 2nd part might need to be in the actual documentation though as I suspect it could be very long. This is part of the release notes so having information on how to use it is very helpful.

**Review by awanlin — COMMENTED** [review]
Hi @jrwpatterson this is pretty awesome and something we had on our backlog to implement so many thanks for this. 

I will note that this may get some push back from the maintainers as I'm not sure they are wanting to change the core Kinds as they can really very from Org to Org. But I think and hope the way you've done this might get it in 🤞

**Comment by jrwpatterson** [comment]
@awanlin the idea I had was keeping the core entity identical so intelliscence and the validators still work for the originals, but the profile can be extended if you so wish for your plugin, I am hoping because from an intellisense and core functionality it works identically and if you add to your profile it will continue to work in all the other places! This just allows organisations to add information that's relevant to then, we want to use it as we want to add slack channels and job titles for certain items. 

Eg a plugin I have written which I might look at adding if this comes about is one where we get the users from slack, and we pull slack channel users in as a way of joining teams through a self service. I want to out the slack direct link on the groups and user. 

So in our case we are a consultancy that has 

- software communities 
- communities for the three major cloud providers 

And a few others! We are pulling the memberships off the slack apis for those communities and then creating groups around them (again so we can have the slack direct link on them) 

That way users that focus on azure see and are owners of the software that is azure related, software community is owner of our software libraries and templates etc etc

Really nice way to self service what groups you are in as our hr system that we are pulling the user details from doesn't have that info!

File: storybook/.storybook/preview-head.html
Comment (jrwpatterson) [reply storybook/.storybook/preview-head.html:3]: That is because to enable any icons to be added I am using the `Icon` from `@material-ui/core/Icon` to enable any icon to be added from webfonts, if there is a better way of doing this happy to change?

File: packages/core-components/src/components/Table/Table.tsx
Comment (jrwpatterson) [reply packages/core-components/src/components/Table/Table.tsx:376]: that is because the TSC at least on my machine was failing to build due to it being `type unknown` and seeing that as an error happy to revert. 
Screen Shot 2022-10-17 at 8 13 08 am

File: .changeset/tender-jeans-clean.md
Comment (jrwpatterson) [reply .changeset/tender-jeans-clean.md:1]: yes sure this might take a while because the adding of catalog users (in my findings) was a little light in the documentation! In fact I reverse engineered the [`Roadie` Okta package](https://github.com/RoadieHQ/roadie-backstage-plugins/tree/main/plugins/backend/catalog-backend-module-okta) to do but I'll see what I can do!

**Comment by jrwpatterson** [comment]
Also I just want to be clear I'm not saying that this way is the correct way to do this! If there is a better way I would love to know it and I can adjust the pr accordingly

File: ADOPTERS.md
Comment (jrwpatterson) [reply ADOPTERS.md:218]: sure I'll pull it

File: .changeset/tender-jeans-clean.md
Comment (jrwpatterson) [reply .changeset/tender-jeans-clean.md:1]: would `https://backstage.io/docs/features/software-catalog/external-integrations` be the place to add this or a new section?

Comment (jrwpatterson) [reply .changeset/tender-jeans-clean.md:1]: done in external-integrations, happy to change as needed.

File: ADOPTERS.md
Comment (awanlin) [reply ADOPTERS.md]: If you make the PR with just this they will merge it in pretty fast!

File: storybook/.storybook/preview-head.html
Comment (awanlin) [reply storybook/.storybook/preview-head.html]: Thanks, no worries 👍

File: packages/core-components/src/components/Table/Table.tsx
Comment (awanlin) [reply packages/core-components/src/components/Table/Table.tsx]: Interesting, I only asked as I was working in that exact file for a PR I have open. I didn't run into any issues running `tsc`. Let's leave this up to the maintainers themselves

File: packages/catalog-model/src/kinds/GroupEntityV1alpha1.ts
Comment (Rugvip) [inline packages/catalog-model/src/kinds/GroupEntityV1alpha1.ts:26]: This addition of opening up the profile with any additional keys and at the same time encoding that the 3 existing ones are special is quite scary. It's likely to lock us into this shape forever unless we ship breaking changes. I'd like to explore other options here where we keep the profile information as it is today. Some related discussion over here as well: #5930

I think a big part of the issue here is that while it's possible to easily customize entity schema, it's not as easy to customize the rendering of things in the frontend. In many places we're simply relying on complete replacements of components right now, but that might change with #11819

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! 👍 

This is a controversial one as I'm not sure we want to make these changes to the entity schema. If you want to break this apart to ship things I'd say the links & docs parts are less controversial and easier to ship.

File: docs/features/software-catalog/external-integrations.md
Comment (jrwpatterson) [reply docs/features/software-catalog/external-integrations.md:361]: yup probably taken from my code where its all codegenned off graphql so all `string | undefined` 😆 will fix...

File: plugins/org/src/components/Cards/Meta/ProfileInfoGroup.tsx
Comment (jrwpatterson) [reply plugins/org/src/components/Cards/Meta/ProfileInfoGroup.tsx:40]: only reason for doing it there is then there are two conditionals purely a small optimisation but too easy either way

File: plugins/org/src/components/Cards/Meta/ProfileInfoGroup.tsx
Comment (jrwpatterson) [reply plugins/org/src/components/Cards/Meta/ProfileInfoGroup.tsx:32]: yup fair enough! sometimes we can over optimise!

File: packages/catalog-model/src/kinds/UserEntityV1alpha1.ts
Comment (jrwpatterson) [reply packages/catalog-model/src/kinds/UserEntityV1alpha1.ts:26]: sure

**Comment by jrwpatterson** [comment]
> Thank you! 👍
> 
> This is a controversial one as I'm not sure we want to make these changes to the entity schema. If you want to break this apart to ship things I'd say the links & docs parts are less controversial and easier to ship.

Hey @Rugvip thats great feedback. I wonder if I change it I think the [Metadata field](https://backstage.io/docs/features/software-catalog/descriptor-format#common-to-all-kinds-the-metadata) I was reading is specifically allowed to be extended, how about instead of extending off the metadata field we do it in there? I would love to have used the `labels` field but that has to be camel case on both the key and the value and some times the value doesn't want to be broken off.. 

But for now I'll probably split this in two the first will be the links with a second pending feedback above for the labels. The links at least would help @JordiPolo with his twin email problem. I'll try and get these changes done in the next few hours.

File: packages/catalog-model/src/kinds/GroupEntityV1alpha1.ts
Comment (jrwpatterson) [reply packages/catalog-model/src/kinds/GroupEntityV1alpha1.ts]: That should be updated now

File: .changeset/tender-jeans-clean.md
Comment (awanlin) [reply .changeset/tender-jeans-clean.md]: Awesome, thanks!

**Comment by jrwpatterson** [comment]
Any updates on this? I'll do a second PR with the profile changes after this for the more controversial discussion! 😆

**Review by Rugvip — APPROVED** [review]
Nice 😁. This is more straightforward, let's :shipit: and discuss the rest separately. Thank you! 👍

**Comment by markyjones** [comment]
This change has become a pain for my team who have a large number of links against our Group entities. 

There was already an `EntityLinksCard` component for displaying links which wrapped long lists into columns which we made good use of. I am struggling to understand why this particular change wasn't either added as an optional feature with a boolean toggle or built to filter on a links `type` [field](https://backstage.io/docs/features/software-catalog/descriptor-format#links-optional) e.g. `contact-info` or similar. 

Given there is no toggle to prevent this feature being enabled I expect we will not be the only teams now forced to make code changes to be able to upgrade. I am not sure if this should be raised as a bug (to add a toggle) or a feature (to add a toggle or filter) in the issues, any direction would be appreciated. 🙇 

Screenshot 2022-11-22 at 09 39 57

**Comment by jrwpatterson** [comment]
I can do a toggle pr in the morning I'm in Brisbane so in about 24 hours

**Comment by andrewthauer** [comment]
That's an interesting point about the links. Why not just use the existing `EntityLinksCard`?

**Comment by suuus** [comment]
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻

**Pull request merged**
